pkg/controller/association/dynamic_watches.go

```diff
@@ -5,84 +5,107 @@
 	// watch the referenced ES cluster for future reconciliations
-	if err := r.watches.ElasticsearchClusters.AddHandler(watches.NamedWatch{
-		Name:    esWatchName(associatedKey),
-		Watched: []types.NamespacedName{esRef},
-		Watcher: associatedKey,
+	if err := ReconcileWatch(associated, associations, r.watches.ElasticsearchClusters, esWatchName(associated), func(association commonv1.Association) types.NamespacedName {
+		return association.AssociationRef().NamespacedName()
 	}); err != nil {
 		return err
 	}
 
 	// watch the user secret in the ES namespace
-	userSecretKey := UserKey(association, esRef.Namespace, r.UserSecretSuffix)
-	if err := r.watches.Secrets.AddHandler(watches.NamedWatch{
-		Name:    esUserWatchName(associatedKey),
-		Watched: []types.NamespacedName{userSecretKey},
-		Watcher: associatedKey,
+	if err := ReconcileWatch(associated, associations, r.watches.Secrets, esUserWatchName(associated), func(association commonv1.Association) types.NamespacedName {
+		return UserKey(association, association.AssociationRef().Namespace, r.UserSecretSuffix)
 	}); err != nil {
 		return err
 	}
 
 	// watch the CA secret in the targeted service namespace
 	// Most of the time it is Elasticsearch, but it could be Kibana in the case of the APMServer
-	if err := r.watches.Secrets.AddHandler(watches.NamedWatch{
-		Name: associatedCAWatchName(associatedKey),
-		Watched: []types.NamespacedName{
-			{
-				Name:      certificates.PublicCertsSecretName(remoteServiceNamer, associationRef.Name),
-				Namespace: associationRef.Namespace,
-			},
-		},
-		Watcher: associatedKey,
+	if err := ReconcileWatch(associated, associations, r.watches.Secrets, associatedCAWatchName(associated), func(association commonv1.Association) types.NamespacedName {
+		ref := association.AssociationRef()
+		return types.NamespacedName{
+			Name:      certificates.PublicCertsSecretName(r.AssociationInfo.AssociatedNamer, ref.Name),
+			Namespace: ref.Namespace,
+		}
 	}); err != nil {
 		return err
 	}
 
+	// set additional watches, in the case of a transitive Elasticsearch reference we must watch the intermediate resource
+	if r.SetDynamicWatches != nil {
+		if err := r.SetDynamicWatches(associated, associations, r.watches); err != nil {
+			return err
+		}
+	}
+
 	return nil
 }
 
+func ReconcileWatch(
+	associated types.NamespacedName,
+	associations []commonv1.Association,
+	dynamicRequest *watches.DynamicEnqueueRequest,
+	watchName string,
+	watchedFunc func(association commonv1.Association) types.NamespacedName,
+) error {
+	if len(associations) == 0 {
+		// clean up if there are none
+		RemoveWatch(dynamicRequest, watchName)
+		return nil
+	}
+
+	toWatch := make([]types.NamespacedName, 0, len(associations))
+	for _, association := range associations {
+		toWatch = append(toWatch, watchedFunc(association))
+	}
+
+	return dynamicRequest.AddHandler(watches.NamedWatch{
```

Comment:
This is is nice and solves all of the regex problems. 👍  It seems so obvious now!

Reply:
Credit goes to @sebgl for suggesting to follow what we do with secure settings 🙂